_posts/2022-08-16-zio-http.md

```diff
@@ -0,0 +1,778 @@
+```
+
+There are four operators to compose these "HTTP applications": `++`, `<>`, `>>>`
+and `<<<`, and the behavior of each is as described from the
+[official documentation](https://dream11.github.io/zio-http/docs/v1.x/dsl/http#composition-of-http-applications).
+
+> ++ is an alias for defaultWith. While using ++, if the first HTTP application
+> returns None the second HTTP application will be evaluated, ignoring the
+> result from the first. If the first HTTP application is failing with a Some[E]
+> the second HTTP application won't be evaluated.
+
+> <> is an alias for orElse. While using <>, if the first HTTP application fails
+> with Some[E], the second HTTP application will be evaluated, ignoring the
+> result from the first. If the first HTTP application returns None, the second
+> HTTP application won't be evaluated.
+
+> `>>>` is an alias for andThen. It runs the first HTTP application and pipes
+> the output into the other.
+
+> <<< is the alias for compose. Compose is similar to andThen. It runs the
+> second HTTP application and pipes the output to the first HTTP application.
+
+### Server
+
+At this pont, we have everything needed to start up an instance of our web
+server:
+
+```scala
+  val program: ZIO[Any, Throwable, ExitCode] = for {
+    _ <- Console.printLine(s"Starting server on http://localhost:$port")
+    _ <- Server.start(port, combined)
+  } yield ExitCode.success
+
+  override def run: ZIO[Any with ZIOAppArgs with Scope, Any, Any] =
+    program
+```
+
+This is a simple entry point, and we only need to give `Server.start` a port
+(defined as `9000` above), and our _composed_ `Http[R, E, Request, Response]`.
+
+Note that `Server.start` internally calls `ZIO.never`, and will block your
+for-comprehension at that point. You should include it last, or append
+`.forkDaemon`, and provide your own logic afterwards.
+
+You can apply some configuration to the `Server` instance, however we won't
+cover this in much capacity in this article. If interested, you can see the
+official documentation
+[here](https://dream11.github.io/zio-http/docs/v1.x/dsl/server/) and
+[here](https://dream11.github.io/zio-http/docs/v1.x/examples/advanced-examples/advanced_server).
+
+## Next Steps
+
+At this stage, we've covered some `zhttp` basics, like pattern matching on a
+requests method and path, to run the appropriate logic. Out next steps will be
+about adding on extra functionality, like CRSF tokens and Authorization, via
+Middleware.
+
+### Middleware
+
+Broadly, the definition of _middleware_ is context dependent; in our realm of
+our discussion, if we're turning a `Request` into a `Response`, then it's
+anything we do _in the middle_ of that process. It may be something behind the
+scenes, like adding logging through the process, or even more intrusive like
+modifying the request to add headers to the request. Middleware is great at
+helping de-couple/re-use business logic. I use the word _intrusive_ to help
```

Comment:
> I use the word _intrusive_ to help emphasize that de-couple doesn't mean make _optional/not required_

Reading the previous sentence, I understand "intrusive" as "tampering with the original request". I see the above tries to give it a different meaning, but it can sow confusion. What do you mean?

Reply:
Yeah, I didn't quite like invasive here - but I was trying to convey that they be a "no-op" , like logging which would not affect downstream logic, or update change behavior, like auth - which could. Perhaps since I used "behind the scenes" for logging, I could keep up the theatrics and use  something like "or even more center-stage".

Reply:
🤔 I think the last part of that paragraph could just be left out - I don't think it provides itself runway to runway to be a valuable discussion